src/common/appState.ts

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {EventEmitter} from 'events';
+
+import {UPDATE_APPSTATE, UPDATE_APPSTATE_TOTALS, UPDATE_APPSTATE_FOR_VIEWID} from 'common/communication';
+import ServerManager from 'common/servers/serverManager';
+import {Logger} from 'common/log';
+
+const log = new Logger('AppState');
+
+export class AppState extends EventEmitter {
+    private expired: Map<string, boolean>;
+    private mentions: Map<string, number>;
+    private unreads: Map<string, boolean>;
+
+    constructor() {
+        super();
+
+        this.expired = new Map();
+        this.mentions = new Map();
+        this.unreads = new Map();
+    }
+
+    updateExpired = (viewId: string, expired: boolean) => {
+        ServerManager.getViewLog(viewId, 'AppState').silly('updateExpired', expired);
+
+        this.unreads.set(viewId, expired);
+        this.emitMentions(viewId);
+    }
+
+    updateMentions = (viewId: string, mentions: number) => {
+        ServerManager.getViewLog(viewId, 'AppState').silly('updateMentions', mentions);
+
+        this.mentions.set(viewId, mentions);
+        this.emitMentions(viewId);
+    };
+
+    updateUnreads = (viewId: string, unreads: boolean) => {
+        ServerManager.getViewLog(viewId, 'AppState').silly('updateUnreads', unreads);
+
+        this.unreads.set(viewId, unreads);
+        this.emitMentions(viewId);
+    };
+
+    clear = (viewId: string) => {
+        ServerManager.getViewLog(viewId, 'AppState').silly('clear');
+
+        this.expired.delete(viewId);
+        this.mentions.delete(viewId);
+        this.unreads.delete(viewId);
+    }
+
+    emitStatus = () => {
+        log.silly('emitStatus');
+
+        this.emit(UPDATE_APPSTATE,
+            this.expired,
+            this.mentions,
+            this.unreads,
+        );
+        this.emit(UPDATE_APPSTATE_TOTALS,
+            [...this.expired.values()].some((value) => value),
+            [...this.mentions.values()].reduce((total, value) => total + value, 0),
+            [...this.unreads.values()].some((value) => value),
+        );
+    };
+
+    private emitMentions = (viewId: string) => {
```

Comment:
Instead of emitMentions, shouldn't be still emitStatus? Since it is for just one viewId (even thought then it calls emitStatus either way) maybe `emitStatusForView`?

Reply:
Yeah that makes sense. Will fix.